### PR TITLE
Add regression test

### LINKED ISSUE #13344: StackOverflow error when compiling some Church-encoded numerals
## Compiler version

`3.0.1`

## Minimized code

The whole example is under [Scastie](https://scastie.scala-lang.org/S7f4vj7pQkKobTuIgWZmBQ), but the problematic code is

```Scala
  type Nat = [A] => (A => A) => A => A

  def Zero: Nat =
    [A] => (f: A => A) => (a: A) => a

  def Succ[A](n: Nat): Nat =
    [A] => (f: A => A) => (a: A) => f(n(f)(a))

  def one: Nat =
    Succ(Zero)

  def mul(m: Nat)(n: Nat): Nat =
    [A] => (f: A => A) => (a: A) => m(n(f))(a)

  def exp(m: Nat)(n: Nat): Nat =
    [A] => (f: A => A) => (a: A) => (n[Nat](mul(m))(one))(f)(a) // AssertionError
    // Adding a type ascription works around the issue
    // [A] => (f: A => A) => (a: A) => (n[Nat](mul(m))(one): Nat)(f)(a) // AssertionError
```

## Output (click arrow to expand)
<details>

```scala
Exception in thread "main" java.lang.AssertionError: assertion failed:  <none> while compiling try/i13344.scala
java.lang.AssertionError: assertion failed:  <none>
        at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
        at dotty.tools.dotc.transform.Erasure$Typer.typedSelect(Erasure.scala:721)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2665)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2757)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2943)
        at dotty.tools.dotc.transform.Erasure$Typer.typedTypeApply(Erasure.scala:815)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2714)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2758)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2943)
        at dotty.tools.dotc.transform.Erasure$Typer.typedApply(Erasure.scala:837)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2695)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2758)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.transform.Erasure$Typer.typedSelect(Erasure.scala:696)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2665)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2757)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2943)
        at dotty.tools.dotc.transform.Erasure$Typer.typedApply(Erasure.scala:837)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2695)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2758)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2943)
        at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1032)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2703)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2758)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2943)
        at dotty.tools.dotc.typer.Typer.$anonfun$34(Typer.scala:2158)
        at dotty.tools.dotc.typer.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:225)
        at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:2158)
        at dotty.tools.dotc.transform.Erasure$Typer.typedDefDef(Erasure.scala:955)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2672)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2757)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2849)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2899)
        at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1053)
        at dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:1026)
        at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1030)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2703)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2758)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2943)
        at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1032)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2703)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2758)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2943)
        at dotty.tools.dotc.typer.Typer.$anonfun$34(Typer.scala:2158)
        at dotty.tools.dotc.typer.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:225)
        at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:2158)
        at dotty.tools.dotc.transform.Erasure$Typer.typedDefDef(Erasure.scala:955)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2672)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2757)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2849)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2899)
        at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1053)
        at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2354)
        at dotty.tools.dotc.transform.Erasure$Typer.typedClassDef(Erasure.scala:1043)
        at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2683)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2687)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2757)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2849)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2899)
        at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1053)
        at dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:1026)
        at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1030)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2703)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2758)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2943)
        at dotty.tools.dotc.typer.Typer.$anonfun$34(Typer.scala:2158)
        at dotty.tools.dotc.typer.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:225)
        at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:2158)
        at dotty.tools.dotc.transform.Erasure$Typer.typedDefDef(Erasure.scala:955)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2672)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2757)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2849)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2899)
        at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1053)
        at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2354)
        at dotty.tools.dotc.transform.Erasure$Typer.typedClassDef(Erasure.scala:1043)
        at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2683)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2687)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2757)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2849)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2899)
        at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1053)
        at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2477)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2728)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2758)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2823)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2827)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2943)
        at dotty.tools.dotc.transform.Erasure.run(Erasure.scala:132)
        at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:308)
        at scala.collection.immutable.List.map(List.scala:246)
        at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:309)
        at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
        at dotty.tools.dotc.Run.runPhases$5(Run.scala:216)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:224)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:231)
        at dotty.tools.dotc.Run.compileSources(Run.scala:166)
        at dotty.tools.dotc.Run.compile(Run.scala:150)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
        at dotty.tools.dotc.Driver.process(Driver.scala:199)
        at dotty.tools.dotc.Driver.process(Driver.scala:167)
        at dotty.tools.dotc.Driver.process(Driver.scala:179)
        at dotty.tools.dotc.Driver.main(Driver.scala:209)
        at dotty.tools.dotc.Main.main(Main.scala)
```
</details>

### PR DESCRIPTION
**Pull request by nicolasstucki**
Closes #13344

The test fails on https://github.com/lampepfl/dotty/releases/tag/3.0.1

### CODE REVIEW

File: tests/pos/i13344.scala [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+  def one: Nat =
+    Succ(Zero)
+
+  def mul(m: Nat)(n: Nat): Nat =
+    [A] => (f: A => A) => (a: A) => m(n(f))(a)
+
+  def exp(m: Nat)(n: Nat): Nat =
+    // the compiler throws a StackOverflowError without the type annotations!
+    [A] => (f: A => A) => (a: A) => (n[Nat](mul(m))(one): Nat)(f)(a)
```
Comment (aumann) [inline tests/pos/i13344.scala:17]: I believe the reproducer has been posted in an unfortunate format -- not sure if you saw the comment
I only get the stack overflow on 3.0.2-RC1 when I change the last line to

`[A] => (f: A => A) => (a: A) => (n(mul(m))(one))(f)(a)`

(that is, when I remove the type annotations at the absolute rhs

**Review by aumann — COMMENTED** [review]
just fyi - not sure if this is an oversight I also ran into - did not test against current nightly, only 3.0.2-RC1

File: tests/pos/i13344.scala (hunk 1 above)
Comment (aumann) [reply tests/pos/i13344.scala:17]: Ooops, I get an AssertionError, not a stack overflow error, sorry.

Comment (smarter) [reply tests/pos/i13344.scala:17]: Indeed, this is still broken on master then.

**Comment by smarter** [comment]
Closing since the issue isn't actually fixed.

**Pull request closed without merging**